internal/server/polling/server.go

```diff
@@ -97,9 +99,75 @@ func (s *Server) poll(ctx context.Context, resource types.NamespacedName, secret
 		return fmt.Errorf("failed to list pull requests: %w", err)
 	}
 
+	// Log the PRs.
 	for _, pr := range prs {
 		s.log.Info("pull request", "pr", pr)
 	}
 
+	labels := map[string]string{
+		bbp.LabelKey: bbp.LabelValue,
+	}
+
+	// List Terraform objects, created by BBP.
+	tfList, err := s.listTerraformObjects(ctx, tf, labels)
+	if err != nil {
+		return fmt.Errorf("failed to list Terraform objects: %w", err)
+	}
+
+	// Create a list of objects to delete.
+	oldTFs := []*infrav1.Terraform{}
+
+	for _, branchTF := range tfList {
```

Comment:
It's not crucial here, but this and the following loop are quadratic -- they both, worst case, look through all the PRs x all the Terraform objects. Can we improve on that? (I am happy for this to be a follow up, since I think this is _correct_ as it stands).

Reply:
How exactly would you like to improve it?

Reply:
I think as a first step, we can merge the two nested loops (delete tfs and create tfs).

For example:
```
// Start with PRs, they are the source of truth.
loop _, pr := prs {

  // Assume we have to create it.
  shouldCreate := true
  
  // Worst case, we iterated through all of them, that means we have to create one.
  loop idx, tf := tfList as tf {
    if pr belongs to tfList {
    
      // We don't care, there is a TF object for this PR, nothing to do.
      tfList[idx] := nil
      
      // We don't have to create it.
      shouldCreate = false
      
      // We can stop.
      break
    }
  }
  
  if shouldCreate {
    createResource(...)
  }
}

for _, tf := tfList {
  // It was not deleted from this list, so there is no related PR.
  if tf != nil {
    delete
  }
}
```

It still has that nested loop, but only once (and not twice as before) and we could remove two extra slices from the code (`oldTFs` and `newPRs`).

It may looks weird, but it can save a lot of cpu cycles and memory.

Reply:
@yitsushi added your optimized loops.

Reply:
If you use a map, you don't need the inner loop:

```go
prSet := map[string]provider.PullRequest
for _, pr := range prs {
    prSet[fmt.Sprintf("%d", pr.Number)] = pr
}

for _, branchTf := range tfList {
    prLabel := branchTf.GetLabels()[LabelPRIDKey]
    if pr, ok := prSet[prLabel]; !ok {
        // no PR for this TF; delete it
        s.deleteTerraform(ctx, branchTf)
    } else {
        // don't need to create a Terraform for this PR, one exists
        delete(prMap, prLabel)
    }
}
for _, pr := range prMap {
    s.createTerraform(ctx, tf, source, pr.HeadBranch, fmt.Sprintf("%d", pr.Number), s.branchPollingInterval); err != nil {
}
```

I left out handling the errors from `create-` and `deleteTerraform`, even so it's a bit shorter I think.

The set does double-duty by keeping track of the set of Terraform objects that should exist and don't. There's a subtle assumption here, which is that at most one Terraform object will have any given PR number (otherwise, all but one will be deleted).

Reply:
Two notes, though:

 1. It would be conventional to "upsert" all resources that are supposed to exist, since objects that already exist might nonetheless need their definitions updated. This simplifies things however you arrange loops. https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/controller/controllerutil has a func specifically for this.
 2. Having tests that demonstrate that it's correct would make it easier to change the particulars of the algorithm with confidence.

Reply:
1. What does upsert means in this context? Do you mean using
https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/controller/controllerutil#CreateOrUpdate

function instead of client.Create?

2.
> Having tests that demonstrate that it's correct would make it easier to change the particulars of the algorithm with confidence.

Do you mean specific tests for upserting objects? Are they different from tests for creating or deleting objects we already discussed?

Reply:
By "upsert" I mean update if it needs updating (say if the original Terraform object had changed part of its spec), or create if it doesn't exist. This is really what the algorithm should do, I think, since you want to keep the branch Terraform objects up to date with respect to the original (and similarly the sources).

I mean tests that the algorithm does what it's supposed to: the equation is something like "given an original Terraform object and its source, for each pull request from the repo there should be a branch Terraform object the same as the original except with planOnly=true, and a source the same as the original source but referring to the pull request branch."

Reply:
Ah, you mean update branch TF objects not based on the PRs (I think no changes should be required for this) but based on the original TF object which might have changed.

So, you would like to add updating branch TFs in a loop too based on the original TF object (like all fields at once with modifying some of them after the update?))? If yes, what function from that package should be used for this?